### PR TITLE
feature : register external packages in the DB

### LINKED ISSUE #1066: [Question] : external packages and post-install hooks
Right now the external packages listed in the `packages.yaml` configuration file return immediately from `Package.do_install`:

``` python
...
        # No installation needed if package is external
        if self.spec.external:
            tty.msg("%s is externally installed in %s" %
                    (self.name, self.spec.external))
            return
...
```

This means that they don't trigger any kind of post-install hook (i.e. no module file gets generated for them) and they don't get registered in the `index.yaml` DB. 

I would like to change the behavior so that external packages are treated mostly like normal packages, in particular : 
- external packages will be added to the DB (recording information on whether they are external or not)
- post-install hooks are triggered for them (and module files get generated)

Can this be of interest for reintegration into `develop`?

### PR DESCRIPTION
**Pull request by alalazo**
##### Modifications
- [x] external packages are registered in the db and generate module files
- [x] added an attribute `external` to each spec in the DB entry
- [x] added unit tests to stress the new feature

fixes #1066

### CODE REVIEW

**Comment by alalazo** [comment]
@tgamblin @becker33 @mathstuf @mplegendre I think this is ready to be reviewed

File: lib/spack/spack/database.py [hunk 1]
```diff
@@ -95,30 +95,33 @@ class InstallRecord(object):
...
+            'explicit': self.explicit,
+            'external': self.external
         }
 
     @classmethod
     def from_dict(cls, spec, dictionary):
         d = dictionary
         return InstallRecord(spec, d['path'], d['installed'], d['ref_count'],
-                             d.get('explicit', False))
+                             d['explicit'], d['external'])
```
Comment (mathstuf) [inline lib/spack/spack/database.py]: Hmm. Seems this would be best as `**dictionary`. Does that work?

File: lib/spack/spack/database.py [hunk 2]
```diff
@@ -307,9 +311,26 @@ def reindex(self, directory_layout):
...
+                        }
+                        self._add(**kwargs)
+                        restored_from_old_data.append(entry.spec)
+                    except Exception:
+                        # Something went wrong, so the spec was not restored
+                        # from old data
+                        pass
+
+                missing_specs = [spec for spec in directory_layout.all_specs(
+                ) if spec not in restored_from_old_data]
```
Comment (mathstuf) [inline lib/spack/spack/database.py]: Why not collect this in the `except` block above?

File: lib/spack/spack/database.py (hunk 1 above)
Comment (alalazo) [inline lib/spack/spack/database.py]: @mathstuf Apparently not... `spec` may be in `dictionary`, so I had to resort to popping it out before doing a `**d`. Would this be good enough ?

File: lib/spack/spack/database.py (hunk 2 above)
Comment (alalazo) [inline lib/spack/spack/database.py]: @mathstuf Good point. Should be solved now

**Comment by tgamblin** [comment]
@alalazo: this looks good to me except for one thing. I think the *spec* should be where things are stored as external or not.   If I build a package with one version of an external package, and then I change the external location, that should be considered a different build and a different package, and it should change the hash.  Would you agree with that?

Places where this would help:
1. This would let you regenerate almost the complete state of the DB.  In the current implementation, `reindex` would not do this b/c the external info is not stored in `spec.yaml`.
1. We would need this information if we're going to use `spec.yaml` descriptors to identify binary packages.  There should be a difference between a package installed with a spack-built version of a spec and one installed with a non-spack-built version.
1. If we don't do this, you could get a hash conflict in the DB if you first build a package with a Spack MPI, then build it with an external MPI.

Right now the `Spec` has an `external` field, but it is only maintained from concretization time to install time.  It's not actually read in from the DB.  I was thinking it would be worth adding, to external specs, a field like this:

```yaml
external:
    path: /path/to/package
    module: name-of-module
```

Either path or module could be specified in the spec.

What do you think?

**Comment by alalazo** [comment]
@tgamblin Do you want this for v0.10 ?

**Comment by alalazo** [comment]
> What do you think?

It seems a good plan to me

**Comment by tgamblin** [comment]
Depends on how long it would take.  If you can do it this week I think it is reasonable to put in.  Imho it's how this should work... what do you think?

**Comment by alalazo** [comment]
> Depends on how long it would take. If you can do it this week I think it is reasonable to put in. Imho it's how this should work... what do you think?

Fine with me. I'll try to do this asap.

**Comment by alalazo** [comment]
@tgamblin Just to let you know: I restarted working on this. The current version should work as you asked, but I didn't have the possibility to check it on Cray (for the `external_module` part). I still need to do polish a bit the code and add tests. Will ping you when done.

**Comment by alalazo** [comment]
@mplegendre @tgamblin PR ready for review, unit tests have been added.

**Comment by adamjstewart** [comment]
@tgamblin A thought I had on changing hashes. Since Spack is no longer compatible with installations done under a previous hashing mechanism, users are forced to either uninstall everything or create a new Spack installation. Could we increment the version number of Spack every time this happens? Instead of incrementing the version during each release, the version of develop should be incremented to reflect changes in the hashing mechanism.

**Comment by alalazo** [comment]
@adamjstewart Just my 2 cents: I don't like very much this idea. I think that as the core of Spack becomes more stable we should try to go for time-based releases (e.g. every `n` months), increment the version number on releases and feature-freeze some time before a new release to focus only on bug-fixes and documentation. 

`develop` shouldn't be treated as a moving stable release because doing so will either prevent experimenting with new design (as we need to be conservative to maintain stability) or it will compromise stability (giving users the impression that the tool is not ready for production). Besides we will be already at Spack `5.0` which would be weird :smile:

**Comment by adamjstewart** [comment]
I agree, and this makes perfect sense for a regular project. But I currently have a `spack-0.9.1` installation, and a `spack-0.10` installation from the last time the hash changed. Once this is merged, I'll have a `spack-0.10.1` installation, but the real core Spack will still be on 0.9.1. We either need to fix Spack so that a change in the hashing algorithm doesn't require you to reinstall it, or we need some way to make it clear that "Spack after this PR is merged" is incompatible with "Spack before this PR is merged". Incrementing the version number is at least one way of indicating this.

**Comment by tgamblin** [comment]
@adamjstewart: you keep saying it's incompatible, but it's not.  It just doesn't reuse the old packages automatically.  That's not the same as"incompatible", and you can link against old stuff with `^/hash`, but I agree that is painful. What I think you really want is something that can make a best effort to reuse what is already there, which is being worked on.  Then it won't matter as much if the hashes change.

Have you looked at @scheibelp's simplified tree PR #2168?  I think that would allow you to provide a consistent view to users right now, regardless of whether the hashes change.

**Comment by tgamblin** [comment]
Something else worth noting: this PR only needs to change the hashes of externals.  It doesn't have to change the hash of anything else.  What kind of change is considered a disruptive hash change?  Right now, if you update the version of zlib, the hashes of things that depend on it will be different after concretization, because they're built with a new version of zlib.  This is arguably *more* disruptive than the change in this PR.  

I think like I said elsewhere, we need another level of indirection, either through a view/profile (like nix) or through trees like #2168.  Both of those would allow you to upgrade *some* things in place, and would hide the hashes behind the indirection layer.  But, because you have the hashes, you have the potential to roll back if you want (as installs are nondestructive), and you can reproduce issues if you need to.

**Comment by alalazo** [comment]
@tgamblin 
> Something else worth noting: this PR only needs to change the hashes of externals.

The implementation here currently does not: each `spec` has a new keyword `external` in the DB which will be
```json
external: false
```
in case of a "regular" package. I can change the PR so that packages that are not external won't have this keyword (and won't change hash), but I think it's clearer to have the keyword always stated explicitly.

**Comment by tgamblin** [comment]
@alalazo: for specs, the implementation currently tries to omit unneeded attributes. This helps reduce the size of the virtual cache (which contains only abstract specs) and the database, and if we ever start round-tripping hand-written specs (e.g. from user environment files or configs), things like external will be omitted in the output if they were omitted in the original input, which I think it more natural. It also makes things like the virtual cache easier to read, as only relevant information is printed for abstract specs.  If everything gets printed things can get kind of long.  It also has the added benefit of reducing hash churn, and it's a little weird to me gon have something that can be a boolean or a path, but that's minor I guess.

Do you think this is the wrong way to go?

**Comment by adamjstewart** [comment]
> What kind of change is considered a disruptive hash change? Right now, if you update the version of zlib, the hashes of things that depend on it will be different after concretization, because they're built with a new version of zlib. This is arguably more disruptive than the change in this PR.

I understand your point. In my mind, a disruptive hash change is when it is no longer possible to specify a particular installation without a hash. This isn't necessarily any more annoying than a new version of zlib being added, but it is a problem for Python packages and `spack activate/deactivate`. We've bounced around the idea of using already installed versions even if a newer version is available in Spack. If we ever actually implement this, how will we handle hash changes?

I guess hash changes aren't as bad as they used to be, or maybe I've just gotten used to them. Once Spack becomes more stable, I hope to do more of my development in a different Spack installation than the one I use for actually installing the software my users depend on. That way I can wipe everything out and start anew more easily, making testing a new package quicker than if Spack had to parse 500+ existing installations.

**Comment by tgamblin** [comment]
> We've bounced around the idea of using already installed versions even if a newer version is available in Spack. If we ever actually implement this, how will we handle hash changes?

Well, to some extent you wouldn't have to.  If we add the ability to concretize based on what is installed, the concretizer would get something like this. 

Say you type `spack install a`, if some `a` is already installed, we might just say "hey that's already installed".   Done.  

If you typed `spack install a@2.7`, and no `a` was installed that satisfies `@2.7`, we'd need to build it.  It will have a new hash, but you expect that b/c it is a new build.

Now suppose and `a` depends on `b@1.0:4.0` when it's `@2.7`:

```
    a@2.7
      ^b@1.0:4.0
```

If some `b /123` is already installed and `Spec('b /123').satisfies('b@1.0:4.0')`, we'd just plug `b` and its dependencies into the DAG:

```
___    a@2.7
123      ^b@2.0.1     ... etc. ...
456        ^c@1.1
789        ^d@4.5
```

Now the `b` subtree is concrete, and we just have to pick a concrete version, variants, compiler, etc. for `a` so that it doesn't conflict with `b /123`.  Once that is done you'll get an `a` with a new hash that depends on the already-installed `b /123` and its dependencies.

Does that make sense?  Basically it's just a choice made during concretization; the new algorithm would avoid rebuilding things in favor of depending on what is there.

If we add a concept of "environments", then maybe you do it like this instead:

```python
01  spack environment create project1
02  spack environment activate project1    # get the project1 env dir in the `PATH`, `MANPATH`, etc.
03  spack install stuff i need for project1
04  spack install more stuff i need for project1
05
06  spack environment create project2
07  spack environment activate project2    # switch to project2
08  spack install stuff i need for project2
09
10  spack environment activate project1    # switch back to project1
11  spack install somethingelse
```

On lines 4 and 11, the `spack install` commands would concretize based only on what is in the `project1` environment, and it would probably only let you install one version of a thing at a time.  Now you can keep separate environments live for different projects.  The issue here is that there is a harder [Version SAT](https://research.swtch.com/version-sat) problem to be solved than what we currently handle.  That's what is being added to the concretizer.

The disadvantage to doing this is that the new combination of packages may not be tested, which is why we'd leave in a deterministic option for users who want it.  That way we could prefer that each package be built with a fully tested combination of dependencies, or we could opt to reuse.  

You could imagine doing something similar to the above for multi-versioned trees instead of for a single environment.  Most HPC centers I know care to a certain extend about combinatorial versions of things, but want other things to be upgraded in place.  We covered that a bit in our [HUST paper on generating RPMs](https://dl.acm.org/citation.cfm?id=3018837) and I think we could do something similar for a tree of packages, but it would be symlinked and backed by a hash tree like the current spack packages are (which would allow you to potentially roll back a deployment).

Anyway, this is getting long.  I would like to solve these problems, though.

File: lib/spack/spack/database.py [hunk 3]
```diff
@@ -62,9 +62,9 @@
 _db_dirname = '.spack-db'
 
 # DB version.  This is stuck in the DB file to track changes in format.
-_db_version = Version('0.9.2')
+_db_version = Version('0.9.3')
 
-# Default timeout for spack database locks is 5 min.
+# Default timeout for spack database locks is 1 min.
```
Comment (tgamblin) [inline lib/spack/spack/database.py]: maybe just remove this and say "timeout for spack database locks in seconds"

File: lib/spack/spack/database.py [hunk 4]
```diff
@@ -372,9 +372,27 @@ def _read_suppress_error():
             old_data = self._data
             try:
                 self._data = {}
+                missing_specs = set(directory_layout.all_specs())
```
Comment (tgamblin) [inline lib/spack/spack/database.py]: why not just build up a set of added specs, and iterate over specs not in it later?  Seems clearer to me than discarding as each spec is installed.

File: lib/spack/spack/database.py [hunk 5]
```diff
@@ -493,7 +511,7 @@ def _add(self, spec, directory_layout=None, explicit=False):
 
         key = spec.dag_hash()
         if key not in self._data:
-            installed = False
+            installed = True if spec.external else False
```
Comment (tgamblin) [inline lib/spack/spack/database.py]: `bool(spec.external)`?

File: lib/spack/spack/hooks/sbang.py [hunk 6]
```diff
@@ -104,8 +104,12 @@ def filter_shebangs_in_directory(directory, filenames=None):
...
 def post_install(pkg):
     """This hook edits scripts so that they call /bin/bash
-       $spack_prefix/bin/sbang instead of something longer than the
-       shebang limit."""
+    $spack_prefix/bin/sbang instead of something longer than the
+    shebang limit.
+    """
+    if pkg.spec.external:
+        tty.debug('SKIP: shebang filtering [external package]')
+        return
```
Comment (tgamblin) [inline lib/spack/spack/hooks/sbang.py]: 👍

File: lib/spack/spack/test/database.py [hunk 7]
```diff
@@ -353,3 +359,19 @@ def fail_while_writing():
     # reload DB and make sure cmake was not written.
     with install_db.read_transaction():
         assert install_db.query('cmake', installed=any) == []
+
+
+def test_external_entries_in_db(database):
+    install_db = database.mock.db
+    rec = install_db.get_record('mpileaks ^zmpi')
```
Comment (tgamblin) [inline lib/spack/spack/test/database.py]: can you put empty lines before these.. just to make it more visually obvious how many cases there are in this test? 😄

File: lib/spack/spack/test/database.py [hunk 8]
```diff
@@ -353,3 +359,19 @@ def fail_while_writing():
...
+    assert rec.spec.external is None
+    assert rec.spec.external_module is None
+    rec = install_db.get_record('externaltool')
```
Comment (tgamblin) [inline lib/spack/spack/test/database.py]: and here

File: lib/spack/spack/test/database.py [hunk 9]
```diff
@@ -353,3 +359,19 @@ def fail_while_writing():
...
+    assert rec.spec.external == '/path/to/external_tool'
+    assert rec.spec.external_module is None
+    assert rec.explicit is False
+    rec.spec.package.do_install(fake=True, explicit=True)
```
Comment (tgamblin) [inline lib/spack/spack/test/database.py]: and here

File: lib/spack/spack/test/database.py (hunk 7 above)
Comment (alalazo) [inline lib/spack/spack/test/database.py]: @tgamblin Sorry for this, but you see I can't help it. Too bad we can't have an automated checker for things like this!  :smile:

File: lib/spack/spack/database.py (hunk 4 above)
Comment (alalazo) [inline lib/spack/spack/database.py]: @tgamblin I am not sure I understood what you meant here. Can you give a look and see if I implemented the thing the way you were asking?

**Comment by alalazo** [comment]
Now with an `etc/spack/packages.yaml` that looks like:
```yaml
packages:
  gettext:
    paths:
      gettext@0.18.3: /usr/lib
    buildable: False
```
I get on `develop`:
```console
$ spack spec -Il petsc%gcc@4.8 arch=linux-ubuntu14-x86_64
Input spec
--------------------------------
     gp5niij  petsc%gcc@4.8 arch=linux-ubuntu14-x86_64 

Normalized
--------------------------------
     6gmqusa  petsc%gcc@4.8 arch=linux-ubuntu14-x86_64 
     tlulk6o      ^blas
     qyk5ol5      ^lapack
     pfxpxnt      ^python@2.6:2.7
     5gd5s7m          ^bzip2
     fvevxql          ^ncurses
     hl37cq7          ^openssl
     ctlbt62              ^zlib
     xry7gic          ^readline
     6o2mnts          ^sqlite

Concretized
--------------------------------
     lhaumeo  petsc@3.7.5%gcc@4.8+boost~complex~debug+double+hdf5+hypre+metis+mpi+mumps+shared+superlu-dist arch=linux-ubuntu14-x86_64 
     hzsu7l5      ^hdf5@1.10.0-patch1%gcc@4.8+cxx~debug+fortran+mpi+pic+shared~szip~threadsafe arch=linux-ubuntu14-x86_64 
     vtqpwar          ^openmpi@2.0.1%gcc@4.8~java~mxm~pmi~psm~psm2~slurm~sqlite3~thread_multiple~tm~verbs+vt arch=linux-ubuntu14-x86_64 
     2bk4mun              ^hwloc@1.11.4%gcc@4.8 arch=linux-ubuntu14-x86_64 
     ulpm33q                  ^libpciaccess@0.13.4%gcc@4.8 arch=linux-ubuntu14-x86_64 
     7a523za                      ^libtool@2.4.6%gcc@4.8 arch=linux-ubuntu14-x86_64 
[+]  cuvfqqc                          ^m4@1.4.17%gcc@4.8+sigsegv arch=linux-ubuntu14-x86_64 
[+]  zyc4d7e                              ^libsigsegv@2.10%gcc@4.8 arch=linux-ubuntu14-x86_64 
[+]  6eyt3nw                      ^pkg-config@0.29.1%gcc@4.8+internal_glib arch=linux-ubuntu14-x86_64 
     c4heqyf                      ^util-macros@1.19.0%gcc@4.8 arch=linux-ubuntu14-x86_64 
[+]  lepicoz          ^zlib@1.2.10%gcc@4.8+pic arch=linux-ubuntu14-x86_64 
     6bnx5pq      ^hypre@2.11.1%gcc@4.8~internal-superlu+shared arch=linux-ubuntu14-x86_64 
     l2wknf2          ^openblas@0.2.19%gcc@4.8~openmp+pic+shared arch=linux-ubuntu14-x86_64 
     7dd2pzs      ^metis@5.1.0%gcc@4.8~debug~gdb~idx64~real64+shared arch=linux-ubuntu14-x86_64 
     uibq42k          ^cmake@3.7.1%gcc@4.8~doc+ncurses+openssl+ownlibs~qt arch=linux-ubuntu14-x86_64 
[+]  o4lmbzi              ^ncurses@6.0%gcc@4.8 arch=linux-ubuntu14-x86_64 
[+]  5y6b5eu              ^openssl@1.0.2j%gcc@4.8 arch=linux-ubuntu14-x86_64 
     pyqzcqu      ^mumps@5.0.2%gcc@4.8+complex+double+float~idx64~metis+mpi~parmetis~ptscotch~scotch+shared arch=linux-ubuntu14-x86_64 
     hs6rnr2          ^netlib-scalapack@2.0.2%gcc@4.8~fpic+shared arch=linux-ubuntu14-x86_64 
     flafiiu      ^parmetis@4.0.3%gcc@4.8~debug~gdb+shared arch=linux-ubuntu14-x86_64 
     a34j7yb      ^python@2.7.13%gcc@4.8~tk~ucs4 arch=linux-ubuntu14-x86_64 
[+]  t34nstu          ^bzip2@1.0.6%gcc@4.8 arch=linux-ubuntu14-x86_64 
[+]  pihnz5p          ^readline@6.3%gcc@4.8 arch=linux-ubuntu14-x86_64 
     slxpiaz          ^sqlite@3.8.5%gcc@4.8 arch=linux-ubuntu14-x86_64 
     l2msugz      ^superlu-dist@5.1.1%gcc@4.8 arch=linux-ubuntu14-x86_64 
```
and on this branch:
```console
$ spack spec -Il petsc%gcc@4.8 arch=linux-ubuntu14-x86_64
Input spec
--------------------------------
     gp5niij  petsc%gcc@4.8 arch=linux-ubuntu14-x86_64 

Normalized
--------------------------------
     6gmqusa  petsc%gcc@4.8 arch=linux-ubuntu14-x86_64 
     tlulk6o      ^blas
     qyk5ol5      ^lapack
     pfxpxnt      ^python@2.6:2.7
     5gd5s7m          ^bzip2
     fvevxql          ^ncurses
     hl37cq7          ^openssl
     ctlbt62              ^zlib
     xry7gic          ^readline
     6o2mnts          ^sqlite

Concretized
--------------------------------
     lhaumeo  petsc@3.7.5%gcc@4.8+boost~complex~debug+double+hdf5+hypre+metis+mpi+mumps+shared+superlu-dist arch=linux-ubuntu14-x86_64 
     hzsu7l5      ^hdf5@1.10.0-patch1%gcc@4.8+cxx~debug+fortran+mpi+pic+shared~szip~threadsafe arch=linux-ubuntu14-x86_64 
     vtqpwar          ^openmpi@2.0.1%gcc@4.8~java~mxm~pmi~psm~psm2~slurm~sqlite3~thread_multiple~tm~verbs+vt arch=linux-ubuntu14-x86_64 
     2bk4mun              ^hwloc@1.11.4%gcc@4.8 arch=linux-ubuntu14-x86_64 
     ulpm33q                  ^libpciaccess@0.13.4%gcc@4.8 arch=linux-ubuntu14-x86_64 
     7a523za                      ^libtool@2.4.6%gcc@4.8 arch=linux-ubuntu14-x86_64 
     cuvfqqc                          ^m4@1.4.17%gcc@4.8+sigsegv arch=linux-ubuntu14-x86_64 
     zyc4d7e                              ^libsigsegv@2.10%gcc@4.8 arch=linux-ubuntu14-x86_64 
     6eyt3nw                      ^pkg-config@0.29.1%gcc@4.8+internal_glib arch=linux-ubuntu14-x86_64 
     c4heqyf                      ^util-macros@1.19.0%gcc@4.8 arch=linux-ubuntu14-x86_64 
     lepicoz          ^zlib@1.2.10%gcc@4.8+pic arch=linux-ubuntu14-x86_64 
     6bnx5pq      ^hypre@2.11.1%gcc@4.8~internal-superlu+shared arch=linux-ubuntu14-x86_64 
     l2wknf2          ^openblas@0.2.19%gcc@4.8~openmp+pic+shared arch=linux-ubuntu14-x86_64 
     7dd2pzs      ^metis@5.1.0%gcc@4.8~debug~gdb~idx64~real64+shared arch=linux-ubuntu14-x86_64 
     uibq42k          ^cmake@3.7.1%gcc@4.8~doc+ncurses+openssl+ownlibs~qt arch=linux-ubuntu14-x86_64 
     o4lmbzi              ^ncurses@6.0%gcc@4.8 arch=linux-ubuntu14-x86_64 
     5y6b5eu              ^openssl@1.0.2j%gcc@4.8 arch=linux-ubuntu14-x86_64 
     pyqzcqu      ^mumps@5.0.2%gcc@4.8+complex+double+float~idx64~metis+mpi~parmetis~ptscotch~scotch+shared arch=linux-ubuntu14-x86_64 
     hs6rnr2          ^netlib-scalapack@2.0.2%gcc@4.8~fpic+shared arch=linux-ubuntu14-x86_64 
     flafiiu      ^parmetis@4.0.3%gcc@4.8~debug~gdb+shared arch=linux-ubuntu14-x86_64 
     a34j7yb      ^python@2.7.13%gcc@4.8~tk~ucs4 arch=linux-ubuntu14-x86_64 
     t34nstu          ^bzip2@1.0.6%gcc@4.8 arch=linux-ubuntu14-x86_64 
     pihnz5p          ^readline@6.3%gcc@4.8 arch=linux-ubuntu14-x86_64 
     slxpiaz          ^sqlite@3.8.5%gcc@4.8 arch=linux-ubuntu14-x86_64 
     l2msugz      ^superlu-dist@5.1.1%gcc@4.8 arch=linux-ubuntu14-x86_64 
```
So: **no hash changes for things that don't involve external packages**. I also checked manually that reindex works with externals:
```console
$ spack find
==> 0 installed packages.

$ spack install --fake gettext zlib
==> gettext@0.18.3 : externally installed in /usr/lib
==> gettext@0.18.3 : generating module file
==> gettext@0.18.3 : registering into DB
==> Installing zlib
==> Building zlib [AutotoolsPackage]
==> Successfully installed zlib
  Fetch: .  Build: 0.04s.  Total: 0.04s.
[+] /home/mculpo/PycharmProjects/spack/opt/spack/linux-ubuntu14-x86_64/gcc-4.8/zlib-1.2.10-lepicozwz5ixytwvfos37z4vz7bobuql

$ md5sum opt/spack/.spack-db/index.json
244e7407ab8d6aecacdb00dbf166d4ca  opt/spack/.spack-db/index.json

$ spack reindex
$ md5sum opt/spack/.spack-db/index.json
244e7407ab8d6aecacdb00dbf166d4ca  opt/spack/.spack-db/index.json
```
Maybe it would be good to add a unit test for this if it is not already there?

**Comment by tgamblin** [comment]
> Maybe it would be good to add a unit test for this if it is not already there?

I think that would be nice but what do we test here?  I suppose you could whether the hash changes when a spec is made external, and whether it has the `external` attribute.  I don't think it's worth the effort to test whether the hash after *this* commit is the same as before this commit unless a spec is external.

**Comment by alalazo** [comment]
@tgamblin I was thinking of a simpler test for a spec/json round-trip if the spec is external. That's what I tested manually in the last lines.

**Comment by tgamblin** [comment]
> I was thinking of a simpler test for a spec/json round-trip if the spec is external. That's what I tested manually in the last lines.

Oh right -- it's late here.  Sigh.

I think that's a really good idea.

**Comment by alalazo** [comment]
@tgamblin the PR is fine from my side

**Comment by alalazo** [comment]
ping

**Comment by alalazo** [comment]
@tgamblin ping

**Comment by krafczyk** [comment]
@alalazo Is this PR still live? I'm finding myself needing something like this.

**Comment by alalazo** [comment]
@krafczyk Yes. I am keeping it up-to-date. I am just waiting for @tgamblin or @becker33 to review it, but it works fine as far as I am concerned.

**Comment by alalazo** [comment]
@tgamblin ping

**Comment by alalazo** [comment]
ping

**Comment by alalazo** [comment]
@tgamblin ping

**Comment by alalazo** [comment]
@tgamblin ping

**Comment by krafczyk** [comment]
@alalazo Are you interested in getting some more changes for this PR? I've got some changes I'd like to be included with this PR in my branch 'fix/external_packages_mods'. Should I submit a PR to your repo?

**Comment by alalazo** [comment]
@krafczyk It depends what the changes are doing :smile: More seriously, I think a PR on the branch is absolutely fine. I'll have time tomorrow to go through it.

**Comment by krafczyk** [comment]
@alalzo, have a look at https://github.com/epfl-scitas/spack/pull/65

**Comment by adamjstewart** [comment]
Ping ping @tgamblin

File: lib/spack/spack/database.py [hunk 10]
```diff
@@ -374,9 +374,29 @@ def _read_suppress_error():
...
+                            'directory_layout': layout,
+                            'explicit': entry.explicit
+                        }
+                        self._add(**kwargs)
+                        processed_specs.add(entry.spec)
+                    except Exception as e:
+                        # Something went wrong, so the spec was not restored
+                        # from old data
+                        tty.debug(e.message)
+                        pass
```
Comment (tgamblin) [inline lib/spack/spack/database.py]: @alalazo: Ok last question on this PR: why does the new code above prefer data in the old DB to data from `spec.yaml` files?  I think the `spec.yaml` files should be considered authoritative.  Is this an artifact of the way this PR started -- with external stored only in the DB and not in the spec?  Could we just keep the old logic?

The logic above is "correct" but I think it does things the wrong way around for reindex.  In particular, reindex is supposed to restore a database from `spec.yaml` files, in case the DB is corrupted or otherwise inaccurate.  

The logic above reads the old DB and adds what it finds *in the old DB file* to the new DB.  It then subtracts the already-read specs from what is available in `spec.yaml` files (via `directory_layout`) and adds the remainder.  We do catch `Exception` above in case anything goes wrong while parsing a spec from the old DB (e.g., if it is corrupted) but this may not be the only thing wrong with the old DB.  The old DB may be corrupted in a valid, readable way, but this would prefer old DB data to data from `spec.yaml` files.  Also, if the hashes in the old DB differ somehow from what is on disk, the algorithm here will perpetuate the old, incorrect hashes, along with any read in from `spec.yaml` files on disk.

Consider also a use case where someone moves their Spack installation and runs `sed` on all the `spec.yaml` files.  I think they should be able to reindex after that and restore a working database, assuming they created valid `spec.yaml` files.  The above logic prevents that.

Basically I think this algorithm should be like it was before: it should loop *first* over `directory_layout.all_specs()` and only *supplement* what it finds in `spec.yaml` with things like `explicit` from the DB.

Finally, note that this doesn't prevent us from finding missing specs.  The `spec.yaml` files store info on the root and all dependencies, and the DB will deduce when deps are missing.

File: lib/spack/spack/test/database.py [hunk 11]
```diff
@@ -86,11 +86,17 @@ def _check_merkleiness():
 
 def _check_db_sanity(install_db):
     """Utiilty function to check db against install layout."""
-    expected = sorted(spack.store.layout.all_specs())
+    pkg_in_layout = sorted(spack.store.layout.all_specs())
     actual = sorted(install_db.query())
 
-    assert len(expected) == len(actual)
-    for e, a in zip(expected, actual):
+    externals = sorted([x for x in actual if x.external is True])
```
Comment (tgamblin) [inline lib/spack/spack/test/database.py]: `if x.external is True` -> `if x.external`

File: lib/spack/spack/test/database.py [hunk 12]
```diff
@@ -86,11 +86,17 @@ def _check_merkleiness():
...
+    nexpected = len(pkg_in_layout) + len(externals)
+
+    assert nexpected == len(actual)
+
+    non_external_in_db = sorted([x for x in actual if x.external is False])
```
Comment (tgamblin) [inline lib/spack/spack/test/database.py]: ` if x.external is False` ->  `if not x.external`

**Comment by alalazo** [comment]
> Is this an artifact of the way this PR started -- with external stored only in the DB and not in the spec? Could we just keep the old logic?

I added the logic to take care of the fact that you may have registered external packages that are not dependencies of anything else (think of an externally installed Intel compiler, or an external licensed application), so that they could be reconstructed as well. I agree that the logic should be reversed, and this is fixed now.

I also added comments in the code to not lose memory of what we said here. Let me kow if I need to rephrase anything to make it clearer.

**Comment by alalazo** [comment]
Ok, this should be ready for another review round.

File: lib/spack/spack/database.py [hunk 13]
```diff
@@ -409,31 +409,83 @@ def _read_suppress_error():
...
-                    str(self._error))
+                    str(self._error)
+                )
                 self._error = None
 
+            # Read first the `spec.yaml` files in the prefixes. They should be
+            # considered authoritative with respect to DB reindexing, as
+            # entries in the DB may be corrupted in a way that still makes
+            # them readable. If we considered DB entries authoritative
+            # instead, we would perpetuate errors over a reindex.
```
Comment (tgamblin) [inline lib/spack/spack/database.py]: 🎉

Comment (tgamblin) [inline lib/spack/spack/database.py]: We *probably* should add an option to propagate only specs from `spec.yaml` files, but that can wait for a separate PR.  I don't think there is an immediate need for it.

File: lib/spack/spack/database.py [hunk 14]
```diff
@@ -409,31 +409,83 @@ def _read_suppress_error():
...
+                            'spec': entry.spec,
+                            'directory_layout': layout,
+                            'explicit': entry.explicit
+                        }
+                        self._add(**kwargs)
+                        processed_specs.add(entry.spec)
+                    except Exception as e:
+                        # Something went wrong, so the spec was not restored
+                        # from old data
+                        tty.debug(e.message)
```
Comment (tgamblin) [inline lib/spack/spack/database.py]: Also for separate PR: it would be good to add something like:

```python
if spack.debug:
    print(traceback.format_exc())
```

So we don't lose the trace.

File: lib/spack/spack/package.py [hunk 15]
```diff
@@ -1081,6 +1081,51 @@ def _stage_and_write_lock(self):
...
+        :param bool explicit: True if the package was requested explicitly by
+            the user, False if it was pulled in as a dependency of an explicit
+            package.
+        """
+        if self.spec.external_module:
+            message = '{s.name}@{s.version} : has external module in {module}'
+            tty.msg(message.format(s=self, module=self.spec.external_module))
+            message = '{s.name}@{s.version} : is actually installed in {path}'
+            tty.msg(message.format(s=self, path=self.spec.external_path))
+        else:
```
Comment (tgamblin) [inline lib/spack/spack/package.py]: generally good style to either `assert self.spec.external` or to say `eilf self.spec.external_path: ... else: raise ValueError("spec was not external")`
